lib/spack/docs/packaging_guide.rst

```diff
@@ -2071,19 +2150,15 @@ The parameters are as follows:
     targets into.  It acts like a string, but it's actually its own
     special type, :py:class:`Prefix <spack.util.prefix.Prefix>`.
 
-``spec`` and ``prefix`` are passed to ``install`` for convenience.
-``spec`` is also available as an attribute on the package
-(``self.spec``), and ``prefix`` is actually an attribute of ``spec``
-(``spec.prefix``).
+The arguments ``spec`` and ``prefix`` are passed only for convenience, as they always
+correspond to ``self.spec`` and ``self.spec.prefix`` respectively.
```

Comment:
Yes, but for commonly used arguments like spec, it would be nice to have easy access to them. This is the reason that the old `Package` install method gave you access to self, spec, and prefix. It's the reason that `url_for_version` gives you self and version, even though version probably equals self.version. If Spack were a Python module for Python developers, I would tend to agree with you. But my hope is that people who are unfamiliar with Python and OO-programming can easily modify or create packages.

At the very least, I think we can agree that we should at least be consistent. Either only pass self to every method, or pass self, spec, and prefix. I vote for the latter obviously. With dependency injection, everyone can get what they want, so it's a good compromise.